datastores/gossip_kv/server/lattices/mod.rs

```diff
@@ -0,0 +1,238 @@
+use std::cmp::Ordering;
+use std::collections::HashSet;
+use std::hash::Hash;
+
+use hydroflow::lattices::{IsBot, IsTop, LatticeOrd, Merge};
+
+/// A bounded set union lattice with a fixed size N.
+///
+/// Once the set reaches size N, it becomes top. The items in the set are no longer tracked to
+/// reclaim associated memory.
+#[derive(Debug, Clone, Eq)]
+
+pub struct BoundedSetLattice<T, const N: usize>
+where
+    T: Eq + Hash,
+{
+    // The set of items in the lattice with invariant:
+    // is_top => items.is_empty() ... i.e. the items are dropped when the lattice reaches top.
+    items: HashSet<T>,
+    is_top: bool,
+}
+
+impl<T, const N: usize> Default for BoundedSetLattice<T, N>
+where
+    T: Eq + Hash,
+{
+    fn default() -> Self {
+        Self {
+            items: HashSet::new(),
+            is_top: N == 0, // This lattice is effectively a unit lattice `()`, if N == 0
+        }
+    }
+}
+
+impl<T> From<()> for BoundedSetLattice<T, 0>
+where
+    T: Eq + Hash,
+{
+    fn from(_: ()) -> Self {
+        Default::default()
+    }
+}
+
+impl<T> From<BoundedSetLattice<T, 0>> for ()
+where
+    T: Eq + Hash,
+{
+    fn from(_: BoundedSetLattice<T, 0>) -> Self {}
+}
+
+impl<T, const N: usize> BoundedSetLattice<T, N>
+where
+    T: Eq + Hash,
+{
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn new_from<U>(items: U) -> Self
+    where
+        U: IntoIterator<Item = T>,
+    {
+        let mut lattice = Self::new();
+        lattice.merge(items);
+        lattice
+    }
+}
+
+impl<T, const N: usize> IsBot for BoundedSetLattice<T, N>
+where
+    T: Eq + Hash,
+{
+    fn is_bot(&self) -> bool {
+        match N {
+            0 => true,
+            _ => self.items.is_empty() && !self.is_top,
+        }
+    }
+}
+
+impl<T, const N: usize> IsTop for BoundedSetLattice<T, N>
+where
+    T: Eq + Hash,
+{
+    fn is_top(&self) -> bool {
+        self.is_top
+    }
+}
+
+impl<T, const N: usize, U> Merge<U> for BoundedSetLattice<T, N>
+where
+    U: IntoIterator<Item = T>,
+    T: Eq + Hash,
+{
+    fn merge(&mut self, other: U) -> bool {
```

Comment:
This would not be commutative, right? I'm not sure if a bounded set is a lattice at all...

Reply:
Oh nevermind I see that you clear the set of items, that works then.